Spreadsheet class + basic code editor functionality

Krihag opened the pull request:
**JS classes**
 - added JS classes for the full spreadsheet and a separate representing each of the different cells.
 - will be used for handling all updates to the spreadsheet but also for storing info like code editor value for each cell
-  added function for creating a new row and/or col for the spreadsheet
- Run function (code editor)
- set/get select methods
- set/get value methods for a specific Cell
- Added the display function into the class to make it simpler to show and update the spreadsheet.
- save/load function for spreadsheed (save function not working) 

**code editor**
- added basic code editor functionality
- user can also access values from the different cells on input from the code editor 
(example A1 = grid[0][0].value, A2 = grid[1][0].value and so on)

**this does NOT have a safe eval functionality yet**
We tried to implement a solution for this, but ended up with issues when trying to use it with code editor. Hopefully this can be fixed soon!


**other changes**
- Changed indexedDB function for storing whole spreadsheet Class instead of only cell data (same with load function)
- Had to remove click event on cell (because turned cell into input) and also since it did not match with design (since it was made before we knew prototype). Reused some of the code and added is as part of logic for "sellect" functionality on spreadsheet class. The arrowkeys functionality for navigating between cells does not currently work because of this though.


(If anyone here has experience with indexedDB please look over the added code for it)

Krihag committed:
added basic functionality for codeEditor on run

not added in HTML, might be better to merge updated version to this first  just add a temp ID in HTML

Krihag committed:
Spreadsheet and Cell class

Krihag committed:
Spreadsheet and Cell class

Krihag committed:
temporary changes for html and main.js

had to comment some functionality out temporary in main.js. Will have to find solution for later

Krihag committed:
temporary changes for html and main.js

had to comment some functionality out temporary in main.js. Will have to find solution for later

Krihag committed:
added temporary functions for storing entire sheet in db.js

theese are just temporary files made by gpt, they should be rewritten by someone that knows indexedDB

Krihag committed:
added temporary functions for storing entire sheet in db.js

theese are just temporary files made by gpt, they should be rewritten by someone that knows indexedDB

Krihag committed:
added values gets passed on initial render (if there is any)

Krihag committed:
added values gets passed on initial render (if there is any)

Krihag committed:
added id for Save changes button and removed default text in textarea

Krihag committed:
added id for Save changes button and removed default text in textarea

Krihag committed:
added temporarily solution for indexedDB with storing whole sheet, but this would need to get rewritten before merge

Krihag committed:
added temporarily solution for indexedDB with storing whole sheet, but this would need to get rewritten before merge

Krihag committed:
passing down allCells to cellRow

Krihag committed:
passing down allCells to cellRow

Krihag committed:
changed logic on createCells (classes) so that if its getting data input from indexedDB it wont create extra arrays

Krihag committed:
changed logic on createCells (classes) so that if its getting data input from indexedDB it wont create extra arrays

Krihag committed:
updated save functionality for Spreadsheet class + the set sellect functionality

Krihag committed:
updated save functionality for Spreadsheet class + the set sellect functionality

Krihag committed:
moved the default functionString removed some unused stuff (that might be added back later)

Krihag committed:
moved the default functionString removed some unused stuff (that might be added back later)

Krihag committed:
rebase / merge confict

Krihag committed:
fixing merge confict / rebase

Krihag committed:
fixed issue with no id on first sellect

Commit:
feat: Add functionality for running code in codeEditor

This commit adds the basic functionality for running code in the codeEditor. It includes a click event handler for the run button, which retrieves the code from the editor and evaluates it. The result is then displayed in the corresponding cell identified by the cellId. Additionally, a safeEval function is introduced to evaluate the code using a Web Worker, ensuring a secure execution environment. This implementation improves the user experience by allowing them to execute their code directly within the codeEditor.

Co-authored-by: Kristian Hagesæter <125972171+Krihag@users.noreply.github.com>

Commit:
Updated HTML to latest verison

Commit:
Updated cell.js file

Krihag committed:
Merge pull request #68 from NoroffFEU/safeEval/codeEditor

Safe eval function

Krihag committed:
merge conflicts

Krihag committed:
fixed issue with loading function if there is no spreadsheet data stored

Krihag committed:
sellectCell border fix for darkmode

Anne-Serine commented on index.html:
```diff
@@ -319,30 +277,27 @@ <h1 class="text-sm flex justify-center p-5 dark:bg-ys-black">
-                      rows="6"
-                    >
-                    </textarea>
+                  <div>
```
I think we need to keep the data-cy attribute for E2E testing with Cypress.
```suggestion
                  <div id="code-wrapper" data-cy="util:code-editor">
```

Krihag replied:
thank you! much appreciated 🥇

Anne-Serine commented on index.html:
```diff
@@ -319,30 +277,27 @@ <h1 class="text-sm flex justify-center p-5 dark:bg-ys-black">
-                        aria-label=""
-                        id=""
-                      />
+                      <div>
+                        <button
+                          type="text"
+                          class="w-20 text-center dark:bg-ys-overlay-15 text-white"
+                          id="selected-cell"
+                        ></button>
+                      </div>
```
Should this be a button? Is this supposed to be clickable? Or just show the value in a styled div?

Krihag replied:
I started a version of it where if you clicked it a list of (up to) last 5 previous cells you had clicked would display (to easier select recently used). But I doubt it will be implemented, and probably not necessary at all. so we could change it back.

Anne-Serine replied:
I see. Sounds good😊

Krihag committed:
changed version number for indexedDB to 2

Krihag committed:
Update index.html

Co-authored-by: Anne-Serine <125912893+Anne-Serine@users.noreply.github.com>

Anne-Serine commented on src/js/spreadsheet/Class/helpers/addCol.js:
```diff
@@ -0,0 +1,31 @@
+  console.log(rows);
+  columnPlacement.append(
+    createEle(
+      'div',
+      'w-28 text-center border-x border-ys-backgroundAndText bg-ys-overlay-15 py-2 snap-start',
```
This should rather be `th` with same classes as the other ´th´s.
```suggestion
      'th',
      'w-28 text-center border-x dark:border-ys-overlay-5 border-ys-amethyst-400 dark:bg-ys-overlay-15 bg-white py-2 snap-start',
```

Anne-Serine commented on src/js/spreadsheet/Class/helpers/addRow.js:
```diff
@@ -0,0 +1,27 @@
+      }),
+    );
+  }
+  const spreadSheetContainer = document.querySelector('.spreadsheet-container');
+
+  // checks first if there is a spreadsheet container to append to (incase of readOnlyMode in future)
+  spreadSheetContainer && spreadSheetContainer.appendChild(cellRow(cols, row));
```
The new table row is being appended outside `<tbody>` in the spreadsheet-container. It should be inside `<tbody>`.

Anne-Serine commented on index.html:
```diff
@@ -218,70 +218,28 @@
-            </li>
+            </div>
           </ul>
+          <button class="default-btn" id="save-changes-btn">
+            Save changes
+          </button>
```
Should we remove this, and rather add the ID to the "quick save" button on the right side panel?
They removed this button from the design...

Krihag committed:
updated suggested changes by Anne-Serine + from her HTML branch.

added changes

Krihag committed:
removed space that was placed wrongly inside ID

Krihag committed:
temporary solution for indexedDB (for storing whole spreadsheet)

Krihag committed:
Changed to auto save sheet when run button for code editor is clicked

also had change some of the code in constructor because of lint error (not supported syntax)

Krihag committed:
Trying to fix emporary solution for indexedDB

Krihag committed:
reverted code fom indexedDB and calling it

New solution was not working, so reverted back

FP22FD commented:
Hi @Krihag  😊 As we discussed in Teams, I would like to help as soon as you fix the merge conflicts.

Krihag committed:
fixed merge conflicts

Krihag commented:
Hey @FP22FD . Thank you very much!
The merge conflict should be resolved now.

Anne-Serine commented on src/js/spreadsheet/Class/helpers/addRow.js:
```diff
@@ -0,0 +1,27 @@
+      }),
+    );
+  }
+  const spreadSheetContainer = document.querySelector('.spreadsheet-container');
```
Rows should be appended to `<tbody>` inside the table. Then the first cell in the appended row will be sticky as well.
```suggestion
  const spreadSheetContainer = document.querySelector('.spreadsheet-container tbody');
```

Krihag replied:
Thank you! 💯

Krihag committed:
Update src/js/spreadsheet/Class/helpers/addRow.js

Co-authored-by: Anne-Serine <125912893+Anne-Serine@users.noreply.github.com>

Krihag committed:
small CSS changes (by suggestion from Anne-Serine)

FP22FD committed:
Restore cell navigation

FP22FD commented:
Hi @Krihag the cell navigation has been restored. You can now connect the cell with the editor, I left some comments to explain how it could be possible using the callbacks that are exposed. If you don't need the callbacks, we can just remove them, everything will keep working.

Krihag commented:
Hey @FP22FD . Great job on it! It looks awesome 💯 
We can just keep the comments there since it might be usefull if a future group takes over this project later. 
I dont think I will do anything more on this project for now atleast.

Voidwaker committed:
Fix: Add checks for cellId retrieval and Web Worker error handling

- Added validation to ensure cellId is retrieved correctly from the h2 element.
- Ensured the element with cellId exists before updating text content to prevent errors.
- Added error handling for Web Worker initialization to avoid runtime issues.
- Updated JSDoc comments for better code readability and maintenance.

Voidwaker committed:
Merge branch 'workflow' into SpreadsheetClass

Voidwaker committed:
Update main.js

Voidwaker committed:
Update main.js

removed duplication

Voidwaker commented:
everything is fixed, if the code is good and up to standard. merge it please

FredrikMohag reviewed (approved):
The cellId validation, Web Worker error handling, and JSDoc comments have all been improved, making the code more robust and easier to maintain

Good job @Voidwaker 👍

FredrikMohag committed:
Merge branch 'workflow' into SpreadsheetClass
